flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/AbstractCastRule.java

```diff
@@ -33,7 +39,11 @@ public CastRulePredicate getPredicateDefinition() {
     }
 
     @Override
-    public boolean canFail() {
-        return false;
+    public boolean canFail(LogicalType inputLogicalType, LogicalType targetLogicalType) {
+        final List<LogicalType> inputFields = LogicalTypeChecks.getFieldTypes(inputLogicalType);
+        final List<LogicalType> targetFields = LogicalTypeChecks.getFieldTypes(targetLogicalType);
+
+        return IntStream.range(0, Math.min(inputFields.size(), targetFields.size()))
```

Comment:
To my understanding this is not correct, don't we need to check all the combinations of each of the inputFields with all the targetFields (nested loop)?

Reply:
So this is a generic logic that is supposed to work with every coimbination of types that has children, that is row to row, map to map, array to array and multiset to multiset. For types without children, it always returns false.

So more than correct i would say this works for these cases, for the others, you need to override this method yourself. Does it makes sense?

Reply:
yep yep, again, I lost the context. all good.
Should there be a check though that both lists have the same size? is there a case that we actually need the `Math.min()`?

Reply:
The thing is that you don't reach this code if the rule is not matched first, hence i don't think we need to be strict here with checks.

Reply:
Just saying, that for someone seeing this code in the future, might be confusing.
If we check for same size, we don't need to do the `Math.min()` as well, and we can recognise some change that breaks this size equality.

Reply:
throw an exception, again, I'm not insisting at all, just as a "safety net".

Reply:
Ok, I modified this code again, now this logic is in `CollectionToCollectionCastRule` and then `RowToRowCastRule` has its custom one. Check it again

Reply:
nice! I like it this way, thank you!